aggregation/src/lib.rs

```diff
@@ -52,45 +53,309 @@ alloy_sol_types::sol! {
+        }
+        match self.0.last() {
+            // If the MerkleMountainRange is empty, push the new peak.
+            None => self.0.push(new_peak),
+            // If the tail subtree is larger, push the new subtree onto the end.
+            Some(peak) if peak.max_depth > new_peak.max_depth => {
+                self.0.push(new_peak);
+            }
+            // If the tail subtree is the same size, combine them and recurse.
+            Some(peak) if peak.max_depth == new_peak.max_depth => {
+                // Will never panic, since we've already checked that there is at least one peak.
+                let peak = self.0.pop().unwrap();
+                self.push_peak(Peak {
+                    digest: commutative_keccak256(&peak.digest, &new_peak.digest),
+                    max_depth: peak.max_depth + 1, //TODO this is fallible
+                })?;
+            }
+            Some(_) => {
+                unreachable!("violation of ordering invariant in Merkle mountain range builder")
+            }
+        };
+        Ok(())
+    }
+
+    /// Finalize the [MerkleMountainRange], combining all peaks into one root. No new nodes can be
+    /// added to a finalized commitment.
+    pub fn finalize(&mut self) -> Result<(), Error> {
+        let root = self.0.iter().rev().fold(None, |root, peak| {
+            Some(match root {
+                Some(root) => commutative_keccak256(&root, &peak.digest),
+                None => peak.digest,
+            })
+        });
+        let Some(root) = root else {
+            return Err(Error::EmptyError);
+        };
+        self.0.clear();
+        self.0.push(Peak {
+            digest: root,
+            max_depth: u8::MAX,
+        });
+        Ok(())
+    }
+
+    /// Finalizes the [MerkleMountainRange] and returns the root, or returns `None` is the
+    /// [MerkleMountainRange] is empty.
+    pub fn finalized_root(mut self) -> Option<Digest> {
+        match self.is_empty() {
+            true => None,
+            false => {
+                // finalize should only fail if the MMR is empty.
+                self.finalize().unwrap();
+                Some(self.0[0].digest)
+            }
+        }
+    }
+
+    /// Returns true if the [MerkleMountainRange] is finalized. No new nodes can be added to a
+    /// finalized [MerkleMountainRange].
+    pub fn is_finalized(&self) -> bool {
+        // If the peak has a max-depth of 255, then the mountain range is finalized and no new
+        // peaks can be pushed to it. Note that this state can only be achieved by calling
+        // `finalize` since it is computationally infeasible to push 2^256 nodes onto the MMR,
+        // although it is theoretically consistent that if an MMR reached a state of being a
+        // single peak with max depth value of 255, it would be naturally finalized.
+        self.0
+            .first()
+            .map(|peak| peak.max_depth == u8::MAX)
+            .unwrap_or(false)
```

Comment:
```suggestion
            .map_or(false, |peak| peak.max_depth == u8::MAX)
```

Shouldn't clippy complain about this?